File: src/main/java/me/vankka/simpleast/core/parser/ParseSpec.java
package me.vankka.simpleast.core.parser;

import me.vankka.simpleast.core.node.Node;

public class ParseSpec<R, T extends Node<R>, S> {
    private T root;
    private boolean isTerminal;
    private S state;
    private int startIndex = 0;
    private int endIndex = 0;

    public ParseSpec(T root, S state, int startIndex, int endIndex) {
        this.root = root;
        this.state = state;
        this.isTerminal = false;
        this.startIndex = startIndex;
        this.endIndex = endIndex;
    }

    private ParseSpec(T root, S state) {
        this.root = root;
        this.state = state;
        this.isTerminal = true;
    }

    public void applyOffset(int offset) {
        startIndex += offset;
        endIndex += offset;
    }

    public T getRoot() {
        return root;
    }

    public boolean isTerminal() {
        return isTerminal;
    }

    public S getState() {
        return state;
    }

    public int getStartIndex() {
        return startIndex;
    }

    public int getEndIndex() {
        return endIndex;
    }

    public static <R, T extends Node<R>, S> ParseSpec<R, T, S> createNonterminal(T node, S state, int startIndex, int endIndex) {
        return new ParseSpec<>(node, state, startIndex, endIndex);
    }

    public static <R, T extends Node<R>, S> ParseSpec<R, T, S> createTerminal(T node, S state) {
        return new ParseSpec<>(node, state);
    }
}

File: src/main/java/dev/vankka/simpleast/core/parser/ParseSpec.kt
package dev.vankka.simpleast.core.parser

import dev.vankka.simpleast.core.node.Node

/**
 * Facilitates fast parsing of the source text.
 *
 *
 * For nonterminal subtrees, the provided root will be added to the main, and text between
 * startIndex (inclusive) and endIndex (exclusive) will continue to be parsed into Nodes and
 * added as children under this root.
 *
 *
 * For terminal subtrees, the root will simply be added to the tree and no additional parsing will
 * take place on the text.
 *
 * @param R The type of render context needed by the node that this contains.
 * @param T The type of node that this contains.
 * @param S The type of state that child nodes will use. This is mainly used to just pass through
 *          the state back to the parser.
 */
class ParseSpec<R, T : Node<R>, S> {
  val root: T?
  val isTerminal: Boolean
  val state: S
  var startIndex: Int = 0
  var endIndex: Int = 0

  constructor(root: T?, state: S, startIndex: Int, endIndex: Int) {
    this.root = root
    this.state = state
    this.isTerminal = false
    this.startIndex = startIndex
    this.endIndex = endIndex
  }

  constructor(root: T?, state: S) {
    this.root = root
    this.state = state
    this.isTerminal = true
  }

  fun applyOffset(offset: Int) {
    startIndex += offset
    endIndex += offset
  }

  companion object {

    @JvmStatic
    fun <R, T : Node<R>, S> createNonterminal(node: T?, state: S, startIndex: Int, endIndex: Int): ParseSpec<R, T, S> {
      return ParseSpec(node, state, startIndex, endIndex)
    }

    @JvmStatic
    fun <R, T : Node<R>, S> createTerminal(node: T?, state: S): ParseSpec<R, T, S> {
      return ParseSpec(node, state)
    }
  }
}

